connector/signaltometrics/config/config.go

```diff
@@ -0,0 +1,337 @@
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/confmap"
+	"go.opentelemetry.io/collector/pdata/pcommon"
+	"go.uber.org/zap"
+)
+
+const (
+	// defaultExponentialHistogramMaxSize is the default maximum number
+	// of buckets per positive or negative number range. 160 buckets
+	// default supports a high-resolution histogram able to cover a
+	// long-tail latency distribution from 1ms to 100s with a relative
+	// error of less than 5%.
+	// Ref: https://opentelemetry.io/docs/specs/otel/metrics/sdk/#base2-exponential-bucket-histogram-aggregation
+	defaultExponentialHistogramMaxSize = 160
+)
+
+var defaultHistogramBuckets = []float64{
+	2, 4, 6, 8, 10, 50, 100, 200, 400, 800, 1000, 1400, 2000, 5000, 10_000, 15_000,
+}
+
+// Config for the connector. The connector can convert all signal types to metrics.
+//
+// For spans, the connector can count the number of span events or aggregate them
+// based on their durations in histograms, summaries, or as 2 sum metrics.
+//
+// For all other event types, the connector can count the number of events.
+//
+// All metrics produced by the connector are in delta temporality.
+type Config struct {
+	Spans      []MetricInfo `mapstructure:"spans"`
+	Datapoints []MetricInfo `mapstructure:"datapoints"`
+	Logs       []MetricInfo `mapstructure:"logs"`
+}
+
+var _ confmap.Unmarshaler = (*Config)(nil)
+
+func (c *Config) Validate() error {
+	if len(c.Spans) == 0 && len(c.Datapoints) == 0 && len(c.Logs) == 0 {
+		return fmt.Errorf("no configuration provided, at least one should be specified")
+	}
+	for _, span := range c.Spans {
+		if err := span.validate(); err != nil {
+			return fmt.Errorf("failed to validate spans configuration: %w", err)
+		}
+	}
+	for _, dp := range c.Datapoints {
+		if err := dp.validate(); err != nil {
+			return fmt.Errorf("failed to validate spans configuration: %w", err)
+		}
+	}
+	for _, log := range c.Logs {
+		if err := log.validate(); err != nil {
+			return fmt.Errorf("failed to validate spans configuration: %w", err)
+		}
+	}
+	return nil
+}
+
+// Unmarshal with custom logic to set default values.
+// This is necessary to ensure that default metrics are
+// not configured if the user has specified any custom metrics.
+func (c *Config) Unmarshal(componentParser *confmap.Conf) error {
+	if componentParser == nil {
+		// Nothing to do if there is no config given.
+		return nil
+	}
+	if err := componentParser.Unmarshal(c, confmap.WithIgnoreUnused()); err != nil {
+		return err
+	}
+	for i, info := range c.Spans {
+		info.ensureDefaults()
+		c.Spans[i] = info
+	}
+	for i, info := range c.Datapoints {
+		info.ensureDefaults()
+		c.Datapoints[i] = info
+	}
+	for i, info := range c.Logs {
+		info.ensureDefaults()
+		c.Logs[i] = info
+	}
+	return nil
+}
+
+// MetricInfo for a data type
+type MetricInfo struct {
+	Name        string `mapstructure:"name"`
+	Description string `mapstructure:"description"`
```

Comment:
Should we have a configurable unit in here too?

Reply:
Hmm, I removed the `unit` (it was present in `spanmetricsv2`) because, unit could be configured with the OTTL converters. Do you see any other use for them?

Reply:
> unit could be configured with the OTTL converters

Do you mean with something like `Milliseconds`? While this converts the value to ms, a user of the metric doesn't know what unit it is in. So I think it would be useful if the unit can be explicitly specified.

Reply:
Right, if we store the unit in Elasticsearch then the metric can be self-describing to the degree that Kibana (or some other client) could visualise it more meaningfully. e.g. if it's "By", then Kibana could autoscale the charts to bytes, megabytes, gigabytes, etc.

Reply:
Ah, I see. So will the unit be added as a dimension?

Reply:
See https://github.com/elastic/opentelemetry-dev/blob/main/docs/design-decisions/ingest/mapping.md#units

I think it would need to be considered a dimension since it would influence grouping.

Reply:
We could always add these dimensions using attributes, something like:

```
signalttometrics:
  attributes:
    - key: unit
       default_value: ms
```

A little bit hacky, but I am a bit reluctant to introduce this because it doesn't directly impact the data manipulation/metric generation and is a bit specific to our (Elastic's) requirement. We could add a unit config and make it a bit more generic with something like:

```
unit:
  enabled: true
  attribute_name: unit
```

but I am not sure overall if we need this.

Reply:
Unit is part of the structured data, it's not a free-form attribute and is not vendor-specific. See https://github.com/open-telemetry/opentelemetry-collector/blob/b06236cc794982916cc956f20828b3e18eb33264/pdata/pmetric/generated_metric.go#L72-L81

I was thinking something like:

```yaml
- name: http.trace.span.duration
      description: Span duration for HTTP spans as a exponential histogram metric
      unit: ms
      attributes:
        - key: http.response.status_code
      exponential_histogram:
        value: Milliseconds(end_time - start_time)
```

Reply:
Unit is already mapped as a dimension in a top-level field. See https://github.com/elastic/elasticsearch/blob/6d6fc66e90da3eddcd9f37d93361b77a6596fed9/x-pack/plugin/otel-data/src/main/resources/component-templates/metrics-otel%40mappings.yaml#L14-L17

Reply:
> Unit is part of the structured data, it's not a free-form attribute and is not vendor-specific

Thanks for the link, I didn't know the OTel model allowed adding unit explicitly 🤦 . Will fix.

Reply:
Fixed in https://github.com/elastic/opentelemetry-collector-components/pull/169/commits/cb1555a18b8e989632da93b7eff19d790771194a